### DISCUSSION TITLE
invalid method calls to method where

### DISCUSSION THREAD

**Discussion by jasperf** [enlightn/enlightn#35] [Q&A] [answered]
We have  a Laravel command based on [Clarkeash Doorman](https://github.com/clarkeash/doorman)'s Invitation package and Enlightn mentioned that line 41 `Invite::where('valid_until', '<', now())` has an invalid method call to method(s): `Your application seems to contain invalid method calls to methods that either do not exist or do not match the method signature or scope.`

Here our command code:

 ```php
<?php

namespace App\Console\Commands;

use Clarkeash\Doorman\Models\Invite;
use Illuminate\Console\Command;

class InviteClean extends Command
{
    /**
     * The name and signature of the console command.
     *
     * @var string
     */
    protected $signature = 'invite:clean';

    /**
     * The console command description.
     *
     * @var string
     */
    protected $description = 'Remove invites that were expired';

    /**
     * Create a new command instance.
     *
     * @return void
     */
    public function __construct()
    {
        parent::__construct();
    }

    /**
     * Execute the console command.
     *
     * @return mixed
     */
    public function handle()
    {
        Invite::where('valid_until', '<', now())
                ->where('uses', 0)
                ->delete();
    }
}
```

I fail to understand the issue here. Would someone please explain what I am missing here?

**Comment by paras-malhotra** [comment]
@jasperf can you run this with the `--details` flag? `Model::where` looks perfectly fine and shouldn't be detected as an error.

**Reply by jasperf** [reply]
Did just update to latest Enlightn and ran with `--details` flag just now and see

```
check 30/61: Your application does not contain invalid method calls. Failed
Your application seems to contain invalid method calls to methods that either do not exist or do not match the method signature or scope.
At /app/Console/Commands/InviteClean.php, line 41: Static call to instance method Illuminate\Database\Eloquent\Builder<Illuminate\Database\Eloquent\Model>::where().
```

**Comment by paras-malhotra** [comment, marked as answer]
Hi @jasperf, the issue is with the `@mixin` docblock annotation [here](https://github.com/clarkeash/doorman/blob/e77575340fa716c504234703c282a49abd793061/src/Models/BaseInvite.php#L17). Here's a [related issue](https://github.com/nunomaduro/larastan/issues/588).

I would suggest that it's best for you to ignore this error using the `ignore_errors` config option. Docs [here](https://www.laravel-enlightn.com/docs/getting-started/configuration.html#ignoring-errors).

**Reply by jasperf** [reply]
So I would need

```php
'ignore_errors' => [
    \Enlightn\Enlightn\Reliability\InvalidMethodCallAnalyzer::class => [
        [
            'path' => 'Illuminate\Database\Eloquent\Builder',
            'details' => '*Static call to instance method*',
        ],
    ],
],
```

in the enlightn config file I assume. Will test this out in a bit.

**Reply by paras-malhotra** [reply]
@jasperf the path would be `app/Console/Commands/InviteClean.php`. You can even auto-generate these by running the `enlightn:baseline` Artisan command to ignore all current errors if you want. Here's the [docs link](https://www.laravel-enlightn.com/docs/getting-started/usage.html#establishing-a-baseline) for the baseline command.

**Reply by jasperf** [reply]
Thanks for that @paras-malhotra . Read at https://github.com/phpDocumentor/fig-standards/issues/35 some more about the usage of `@mixin` to make the IDE (PHP Storm mainly?) load/recognize certain classes the IDE has issues with so you can see reference more properly. But now it seems Enlightn / PHPStan can't work properly with `@mixin Builder` set up so we need to exclude the while file from checks.

I would think a PHPStan fix like by adding this to `phpstan.neon`:

```php
parameters:
    mixinExcludeClasses:
        - Illuminate\Database\Eloquent\Builder
```
See [PHPStan docs](https://phpstan.org/config-reference)


as shown at https://github.com/nunomaduro/larastan/issues/588 would be better then.. Or at least one line per `@mixin` reference perhaps. I mean, you would not need to exclude all the files that benefit from using `@mixin`.. or am I seeing this wrong here?

**Reply by jasperf** [reply]
Come to think of it.. would be better if we could somehow just ignore `@mixin` and still have the option to check files that refer to files that use... `@mixin`.

In my App Larastan (perhaps to work with mixins in Larastan and in other apps just to use the mixins)  many files have `@mixin`:

```
...
./vendor/clarkeash/doorman/src/Models/BaseInvite.php: * @mixin Builder
./vendor/nunomaduro/larastan/stubs/Redis/Connection.stub: * @mixin \Redis
./vendor/nunomaduro/larastan/stubs/Contracts/Pagination.stub: * @mixin \Illuminate\Support\Collection
./vendor/nunomaduro/larastan/stubs/Contracts/Pagination.stub: * @mixin \Illuminate\Pagination\Paginator
./vendor/nunomaduro/larastan/stubs/Contracts/Pagination.stub: * @mixin \Illuminate\Support\Collection
./vendor/nunomaduro/larastan/stubs/Contracts/Pagination.stub: * @mixin \Illuminate\Pagination\LengthAwarePaginator
./vendor/nunomaduro/larastan/stubs/Logger.stub:     * @mixin \Illuminate\Log\LogManager
./vendor/nunomaduro/larastan/stubs/Logger.stub:     * @mixin \Monolog\Logger
./vendor/nunomaduro/larastan/stubs/Facades.stub: * @mixin \Redis
./vendor/nunomaduro/larastan/stubs/Facades.stub: * @mixin \Illuminate\Redis\RedisManager
./vendor/nunomaduro/larastan/stubs/Facades.stub: * @mixin \Illuminate\Contracts\Redis\Factory
./vendor/nunomaduro/larastan/stubs/Facades.stub: * @mixin \Illuminate\Database\DatabaseManager
./vendor/nunomaduro/larastan/stubs/Facades.stub: * @mixin \Illuminate\Database\Connection
./vendor/nunomaduro/larastan/stubs/Facades.stub: * @mixin \Illuminate\Database\ConnectionInterface
./vendor/nunomaduro/larastan/stubs/Facades.stub: * @mixin \Illuminate\Queue\QueueManager
./vendor/nunomaduro/larastan/stubs/Facades.stub: * @mixin \Illuminate\Queue\Queue
./vendor/nunomaduro/larastan/stubs/JsonResource.stub: * @mixin TModelClass
./vendor/nunomaduro/larastan/stubs/Pagination.stub: * @mixin \Illuminate\Support\Collection
Binary file ./vendor/bin/phpstan.phar matches
./vendor/ctf0/media-manager/src/resources/assets/sass/partials/vars.scss:@mixin media($amount, $point) {
./vendor/webmozart/assert/CHANGELOG.md:This interface is linked to the `Assert` class with a `@mixin` annotation. Most IDE's have supported this
./vendor/webmozart/assert/src/Assert.php: * @mixin Mixin
./vendor/laravel/framework/src/Illuminate/Database/DatabaseManager.php: * @mixin \Illuminate\Database\Connection
....
./vendor/laravel/framework/src/Illuminate/Filesystem/FilesystemManager.php: * @mixin \Illuminate\Contracts\Filesystem\Filesystem
./vendor/laravel/framework/src/Illuminate/Filesystem/FilesystemAdapter.php: * @mixin \League\Flysystem\FilesystemInterface
./vendor/laravel/framework/src/Illuminate/Testing/TestResponse.php: * @mixin \Illuminate\Http\Response
./vendor/laravel/framework/src/Illuminate/Support/ViewErrorBag.php: * @mixin \Illuminate\Contracts\Support\MessageBag
./vendor/laravel/framework/src/Illuminate/Queue/Capsule/Manager.php: * @mixin \Illuminate\Queue\QueueManager
./vendor/laravel/framework/src/Illuminate/Queue/Capsule/Manager.php: * @mixin \Illuminate\Contracts\Queue\Queue
./vendor/laravel/framework/src/Illuminate/Queue/QueueManager.php: * @mixin \Illuminate\Contracts\Queue\Queue
./vendor/laravel/framework/src/Illuminate/Routing/Router.php: * @mixin \Illuminate\Routing\RouteRegistrar
./vendor/laravel/framework/src/Illuminate/Collections/HigherOrderCollectionProxy.php: * @mixin \Illuminate\Support\Enumerable
./vendor/laravel/framework/src/Illuminate/Collections/HigherOrderWhenProxy.php: * @mixin \Illuminate\Support\Enumerable
..
```

See https://github.com/laravel/framework/pull/20229

**Reply by jasperf** [reply]
Well as it seems mainly Illuminate classes have this `@mixin` only and few others I will try to use

```php
'ignore_errors' => [
    \Enlightn\Enlightn\Reliability\InvalidMethodCallAnalyzer::class => [
        [
            'path' => 'app/Console/Commands/InviteClean.php',
            'details' => '*Static call to instance method*',
        ],
    ],
]
```

but still saw _Your application seems to contain invalid method calls to methods that either do not exist or do not match the method signature or scope._ `At /app/Console/Commands/InviteClean.php, line 41: Static call to instance method Illuminate\Database\Eloquent\Builder<Illuminate\Database\Eloquent\Model>::where().`

**Update**: had to `php artisan config:clear` My bad.

**Reply by paras-malhotra** [reply]
Hi @jasperf, try auto-generating the `ignore_errors` using `php artisan enlightn:baseline`. You can copy the `ignore_errors` entry from there I think.
